### PR TITLE
chore(package): update `uglifyjs-webpack-plugin` v0.4.6...v1.1.1

### LINKED ISSUE #992: An error in UglifyJSPlugin when using html-loader
I use webpack and UglifyJSPlugin to build my backbone application. When building this application an error arises.

```
ERROR in ./~/html-loader!./templates/messageTemplate.html
Module build failed:
 @ ./js/message-view.js 1:22-72
```

I have two undersore's templates:

``` html
<!-- first template -->
<label><%- prompt %></label>
```

``` html
<!-- second template -->
<div <%= color ? ('style="color:' + color + '"') : '' %>>
    <%= link ? ('<a href="' + link + '">') : '' %>
        <%= message %>
    <%= link ? ('</a>') : '' %>
</div>
```

Only the second template have an error.

I would like to know whether I need to write a configuration in UglifyJSPlugin to fix the problem.

Thank you!

### LINKED ISSUE #2023: UglifyJsPlugin is corrupting some Bootstrap CSS
In my project, I'm using a simple Bootstrap dropdown, as described here:

http://getbootstrap.com/javascript/#dropdowns

Here's the markup (with React properties removed):

```
<div class="dropdown">
  <button class="btn btn-default dropdown-toggle" data-toggle="dropdown">
    <span>Info</span>
    <span class="caret"></span>
  </button>
  <ul class="dropdown-menu">
    <li><a href="#" data-value="critical">Critical</a></li>
    <li><a href="#" data-value="error">Error</a></li>
    <li><a href="#" data-value="warning">Warning</a></li>
    <li><a href="#" data-value="info">Info</a></li>
  </ul>
</div>
```

When I enable the Uglify plugin, the caret inside the button no longer appears. In the styles in my Chrome dev tools, I can see that the following style is getting removed by Uglify:

`border-top: 4px dashed;`

And that causes the caret to not appear.

I've tried setting all the compressor options to false, but the problem persists. I've tried disabling minification on my CSS loader, but that doesn't have any effect either. The problem seems to be with UglifyJS.

Any ideas here? I'm not sure how to proceed with debugging.

### LINKED ISSUE #3673: Consider moving `UglifyJsPlugin` out of main package.
Relative to other plugins, it has a disproportionate number of dependencies.  Additionally, with the advent of [`babili`](https://github.com/babel/babili), `uglify` is no longer a necessary part of a production build process.

### LINKED ISSUE #4910: Upgrade uglify to v3
**Do you want to request a *feature* or report a *bug*?**

Feature

**If this is a feature request, what is motivation or use case for changing the behavior?**

Upgrade uglify to v3. Since it is a breaking change to options that are passed directly to uglify, it would be best done as part of a webpack breaking change, e.g. v3

### PR DESCRIPTION
**Pull request by michael-ciniawsky**
### `Issues`

- Fixes #992
- Fixes #2023
- Fixes #3673
- Fixes #4910

**What kind of change does this PR introduce?**

- [x] Chore

**Did you add tests for your changes?**

- [x] Yes

**If relevant, link to documentation update:**

- [ ] TODO

**Explain the motivation for making this change. What existing problem does the pull request solve**

Support for ES2015+ Minification, Caching && Parallelization

**Does this PR introduce a breaking change?**

- [x] Yes

### CODE REVIEW

**Comment by michael-ciniawsky** [comment]
@sokra What is the recommended way to update `expected.txt` ?

**Comment by sokra** [comment]
Looks like the license extracting feature is missing from the uglify-js-webpack-plugin.

**Comment by michael-ciniawsky** [comment]
1) `exclude` not working? (maybe a bug `uglifyjs-plugin`)

2) `// this test is based off https://github.com/mishoo/UglifyJS2/blob/master/test/compress/screw-ie8.js`, but seems to be removed from `uglify-es` (?)

```diff
+ t.exports=function(t){return function(n){try{t()}catch(t){n(t)}}}
- function r(t){return function(n){try{t()}catch(t){n(t)}}}
```

3)

```diff
- /** @preserve comment should be extracted extract-test.1 */

var foo = {};

+ /** @preserve comment should be extracted extract-test.1 */
```

Would 'fix' it, but it's a scoping issue (?)...

4)

```diff
+ o.exports=function(){console.log(4),console.log(6),console.log(4),console.log(7)}
- function e(){var o=2;o=3,console.log(1+o),console.log(o+3),console.log(4),console.log(1+o+3)}
```

**Comment by sokra** [comment]
2, 3, 4 looks fine. What's the progress with 1?

**Comment by michael-ciniawsky** [comment]
https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/164 should fix 1) but I'm waiting for a release and revisit this PR then :)

**Comment by sokra** [comment]
great

**Comment by evilebottnawi** [comment]
@sokra @michael-ciniawsky seems we have problem(s) on windows 10, i think we need solve this before merge https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/118, I'll test it today

**Comment by evilebottnawi** [comment]
Strange:
```shell

 PASS  test\cache-options.test.js
 › 6 snapshots written.
 › 6 obsolete snapshots found.
 FAIL  test\all-options.test.js
  ● when applied with all options › matches snapshot

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot 1.

    - Snapshot
    + Received

      Array [
    -   "Error: main.0c220ec66316af2c1b24.js from UglifyJs
    +   "Error: main.abd5b0696a29aaef3da3.js from UglifyJs
      Dropping unused variable a [./test/fixtures/entry.js:4,0]",
      ]

      at test/all-options.test.js:71:24
          at <anonymous>

 › 2 snapshots written.
 › 2 obsolete snapshots found.
 › 1 snapshot test failed.
 PASS  test\parallel-options.test.js
 › 6 snapshots written.
 › 6 obsolete snapshots found.
 PASS  test\empty-options.test.js
 › 2 snapshots written.
 › 2 obsolete snapshots found.
 PASS  test\ecma.test.js
 FAIL  test\invalid-options.test.js
  ● when applied with invalid options › matches snapshot

    expect(value).toMatchSnapshot()

    Received value does not match stored snapshot 1.

    - Snapshot
    + Received

      Array [
    -   "Error: main.0c220ec66316af2c1b24.js from UglifyJs
    +   "Error: main.abd5b0696a29aaef3da3.js from UglifyJs
      DefaultsError: `invalid-option` is not a supported option",
    -   "Error: manifest.6afe1bc6685e9ab36c1c.js from UglifyJs
    +   "Error: manifest.813fd895e8457b302f41.js from UglifyJs
      DefaultsError: `invalid-option` is not a supported option",
      ]

      at test/invalid-options.test.js:24:22
          at <anonymous>

 › 2 snapshots written.
 › 2 obsolete snapshots found.
 › 1 snapshot test failed.
 PASS  test\extract-comments-options.test.js
 PASS  test\extract-option-set-to-a-single-file.test.js
 PASS  test\uglify\worker.test.js
 PASS  test\supports-multicompiler.test.js
 › 3 snapshots written.
 › 3 obsolete snapshots found.
 PASS  test\exclude-option.test.js
 › 8 snapshots written.
 › 8 obsolete snapshots found.
 PASS  test\uglify\serialization.test.js
 PASS  test\UglifyJsPlugin.test.js

Snapshot Summary
 › 29 snapshots written in 7 test suites.
 › 2 snapshot tests failed in 2 test suites. Inspect your code changes or run with `npm test -- -u` to update them.
 › 29 obsolete snapshots found, run with `npm test -- -u` to remove them.

Test Suites: 2 failed, 11 passed, 13 total
Tests:       2 failed, 104 passed, 106 total
Snapshots:   2 failed, 29 added, 61 passed, 92 total
Time:        12.297s
Ran all test suites.
```

**Comment by evilebottnawi** [comment]
any ideas why hash is difference?

**Comment by michael-ciniawsky** [comment]
@evilebottnawi I don't have a windows machine/box available 😅, the hashes 'only' changed when running the tests on windows or did this happen in general ?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky on windows

**Comment by michael-ciniawsky** [comment]
Comment extracting is failing again 😞, we need to investigate why (maybe related to [#168](https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/168), but unlikely) or I'm using the wrong options 

@evilebottnawi could you take a look aswell please when you have time ? 🙃

**Comment by evilebottnawi** [comment]
@michael-ciniawsky yep, today

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
1. WIP
2. Need rewrite test, we use 
```js
new webpack.optimize.UglifyJsPlugin({
    uglifyOptions: {
        output: {
            comments: true
        }
    },
    extractComments: true,
    include: [
        "extract.js"
    ]
})
```

From documentation `output.comments: true`:
> comments (default false) -- pass true or "all" to preserve all comments, "some" to preserve some comments, a regular expression string (e.g. /^!/) or a function

We pass `true` and our `extract.js` contains all comments. We should pass `false` to remove all comments.

**Comment by michael-ciniawsky** [comment]
Any progress on this, I will take a look later when time but want to check if someone already has something in mind/is working on a solution ?

**Comment by evilebottnawi** [comment]
@michael-ciniawsky right now working and thinking

**Comment by evilebottnawi** [comment]
@michael-ciniawsky we need rewrite test
```js
new webpack.optimize.UglifyJsPlugin({
	uglifyOptions: {
		output: {
			comments: true
		}
	},
	extractComments: true,
	include: [
		"extract.js"
	]
})
```
to
```js
new webpack.optimize.UglifyJsPlugin({
	uglifyOptions: {
		output: {
			comments: false
		}
	},
	extractComments: /^\**!|@preserve|@license|@cc_on/,
	include: [
		"extract.js"
	]
})
```
Why?
1. `comments: false` we don't want comments in `extract.js`.
2. `extractComments: /^\**!|@preserve|@license|@cc_on/` we want only license and etc comment to extract
3. We have only one failed assert `source.should.containEql("comment should be extracted extract-test.1");` in `should extract comments to separate file` because `uglify-es` don't pass `"@preserve", "@license" or "@cc_on"` and etc comments in function (i think it is bug in uglify), i will create issue on this.

What do you think?

**Comment by michael-ciniawsky** [comment]
@evilebottnawi These changes mostly fixes it, but 

1. `comments: false`  should be the default according to `uglify-es` docs, but tests fail when I remove `uglifyjsOptions` entirely, so either it's a `uglify-es` bug or `uglifyjs-webpack-plugin` overrides the defaults in e.g `buildDefaultOptions` ?

2. I assumed this should also be the default  `extractComments: true === /^\**!|@preserve|@license|@cc_on/` ? At least for `/(\/**!|@preserve|@license)/` according to the docs ? 

I still need to move the the first comment under the var declaration as it seems, otherwise it won't be extracted (?) 

**extract.js**
```diff
- /** @preserve comment should be extracted extract-test.1 */

var foo = {};

+ /** @preserve comment should be extracted extract-test.1 */
```

Which maybe a [scope] bug/change between `uglify-js` <=> `uglify-es`, I'm not sure tbh

**Comment by evilebottnawi** [comment]
Issue around `@preserve` https://github.com/mishoo/UglifyJS2/issues/2500

**Comment by evilebottnawi** [comment]
@michael-ciniawsky 
1. Seems bug in `uglifyjs-webpack-plugin` because we don't directly pass `false` in `uglify-es`, WIP on this.
2. `extractComments: true` extract all comments, but in docs we have https://github.com/webpack-contrib/uglifyjs-webpack-plugin#boolean-2, seems also bug. Should we interpreted `true` as `/^\**!|@preserve|@license|@cc_on/`? It is easy bug fix

**Comment by michael-ciniawsky** [comment]
1. 👍 
2. I would personally say so, but if other opinions exists I won't object

**Comment by michael-ciniawsky** [comment]
ℹ️ ⚠️ Still blocked by at least [#173](https://github.com/webpack-contrib/uglifyjs-webpack-plugin/pull/173) (Windows IPC Bugfix), awaiting release

File: test/configCases/plugins/uglifyjs-plugin/extract.js [hunk 1]
```diff
@@ -1,7 +1,9 @@
-/** @preserve comment should be extracted extract-test.1 */
-
 var foo = {};
 
+// ⚠️ move the following comment back to the top
```
Comment (sokra) [inline test/configCases/plugins/uglifyjs-plugin/extract.js]: Is this still relevant?

**Comment by sokra** [comment]
@michael-ciniawsky Should I merge it or do you still want to change something?

**Comment by michael-ciniawsky** [comment]
Should be g2g 👍 [mishoo/UglifyJS2#2500](https://github.com/mishoo/UglifyJS2/issues/2500) is the only known issue atm, but we can't do anything about it within the scope `uglifyjs-webpack-plugin`.

File: test/configCases/plugins/uglifyjs-plugin/extract.js (hunk 1 above)
Comment (michael-ciniawsky) [reply test/configCases/plugins/uglifyjs-plugin/extract.js]: Yep, but it's seems to be a known upstream issue in `uglify-es`

**Comment by sokra** [comment]
Thanks

**Pull request merged**